subprojects/ide-native/src/main/java/org/gradle/ide/xcode/plugins/XcodePlugin.java

```diff
@@ -337,33 +343,21 @@ private static String toFileType(PBXTarget.ProductType productType) {
         }
     }
 
-    private static boolean isRoot(Project project) {
-        return project.getParent() == null;
-    }
-
-    private static XcodeExtension xcodeModelFor(Project project) {
-        return project.getExtensions().getByType(XcodeExtension.class);
-    }
-
-    private void registerXcodeProjectArtifact(Project project, GenerateXcodeProjectFileTask projectTask) {
-        ProjectLocalComponentProvider projectComponentProvider = ((ProjectInternal) project).getServices().get(ProjectLocalComponentProvider.class);
-        ProjectComponentIdentifier projectId = newProjectId(project);
-        projectComponentProvider.registerAdditionalArtifact(projectId, createXcodeProjectArtifact(projectId, xcodeModelFor(project), projectTask));
-    }
-
-    private static LocalComponentArtifactMetadata createXcodeProjectArtifact(ProjectComponentIdentifier projectId, XcodeExtension xcode, Task projectTask) {
-        XcodeProject xcodeProject = xcode.getProject();
-        PublishArtifact publishArtifact = new XcodeProjectArtifact(xcodeProject, projectTask);
-        return new PublishArtifactLocalArtifactMetadata(projectId, publishArtifact);
+    private static PublishArtifact createXcodeArtifact(Project project) {
+        DefaultXcodeProject xcodeProject = ((DefaultXcodeExtension)project.getExtensions().getByType(XcodeExtension.class)).getProject();
+        Task byName = project.getTasks().getByName("xcodeProject");
+        return new XcodeProjectArtifact(xcodeProject, byName);
     }
 
-    private void addIncludedBuildToWorkspace(final Project project, GenerateXcodeWorkspaceFileTask workspaceTask) {
-        final ServiceRegistry serviceRegistry = ((ProjectInternal) project).getServices();
+    private void configureWorkspace(final Project project, GenerateXcodeWorkspaceFileTask workspaceTask) {
         workspaceTask.dependsOn(new Callable<List<TaskDependency>>() {
             @Override
             public List<TaskDependency> call() throws Exception {
+                // Force external source dependencies project evaluation
+                project.getConfigurations().getByName("ide").getResolvedConfiguration();
+
                 return CollectionUtils.collect(
-                    allXcodeprojArtifactsInComposite(serviceRegistry),
```

Comment:
We still want to do what we've previously done, rather than use a configuration to trigger the resolution of some source dependency builds, otherwise we're only going to include the projects that are depended on by the root project and miss everything that is not. For example, if the root project is just a container that has no components, then the workspace will basically be empty.

I would not change how we decide which project files to include in the workspace at this point. For now, we only want to change how we generate the project files.

Instead, we want to change how the project file is generated, to use as input a filtered view of the compile include path/link files/import modules. To do this, we take the configuration, an use `getIncoming().artifactView(action)`. We configure the view (in the action) using the `componentFilter()` method to apply a component filter. This filter would ignore those projects that are going to be included as an Xcode project and accept those that are going to be included as binaries.

Reply:
I'm not sure I completely understand. We aren't changing how we include project files in the workspace. The `getIdeArtifactMetadata` is the same implementation of what used to be `allXcodeprojArtifactsInComposite`. The reason why we are forcing the resolution is because `workspace.dependsOn` will not find the source dependency project as they weren't evaluated yet. This means they are not part of the `ProjectPathRegistry`. However when the task executes and resolve `workspaceTask.setXcodeProjectLocations`, the source dependency project have been evaluated and now we find the Xcode project path. This cause the workspace to include Xcode project that will not be built because we don't have any task dependency on them.

You are right that we could add a dependency from Xcode project generation task to binary compile configuration. This would, in fact, fix some of the problems, maybe all of them, that we have now.

Reply:
We are in fact changing the behaviour by introducing the `ide` configuration. Resolving this before calculating the set of projects will cause some (but not all necessarily) of the source dependency builds to become visible. We should just not use the `ide` configuration.

Reply:
> You are right that we could add a dependency from Xcode project generation task to binary compile configuration

We have to do this, otherwise the result will be missing stuff. These tasks are working by coincidence at the moment.

There are several different things we need to resolve here:

- The set of headers to make visible in a project file.
- The set of swift module files to make visible in a project file.
- The set of link files to make visible in a project file (or maybe not, if Xcode doesn't need these for any kind of content completion or navigation).
- The set of project files to make visible in a workspace file.

Each of these should end up working in a similar way. That is, for each the sets of files, there's a lazy `FileCollection` to be resolved, and when resolved contains the files that are required and which also carries the task dependency information required to build those files. The `FileCollection` should be an input to the task that generates the consuming xcode file, so that all of the relevant source dependencies are included and all of the relevant files generated when the generation task runs. So, the project generation task should take as input the `FileCollection` containing the headers and the `FileCollection` containing the swift modules for that project. And the workspace generation task should take as input the `FileCollection` that contains the set of project files to include in the workspace.

These files should all be filtered in the same way, which is to take the component identifier of each component in the result, and throw away some of the project components. For the headers and swift module files (and link files if required), we throw away components that are from projects in explicitly included builds. For the projects to make visible in the workspace, we throw away components that are not from projects in explicitly included builds. In both cases, we ask the build registry whether a project is from an explicitly included build or not, adding the relevant methods and data to the registry in order to ask the question.

The main difference between the headers and swift modules to make visible in a project, and the projects to make visible in a workspace, is where the _input_ dependencies for the files to be resolved come from. For the headers, we use the C++ compile configuration. For the swift modules, we use the Swift compile configuration. And for the projects, we query the registry for the set of all known projects. Other than that, thing should be pretty much the same.

I would start with the headers and swift modules, get that working, and then come back in another PR to sort out the project files.